docs/resources/server/loottables/datagen.md

```diff
@@ -0,0 +1,200 @@
+- `ScoreboardValue#fromScoreboard`: Given an entity target, a score name and (optionally) a scale value, retrieves the given scoreboard value for the entity target, multiplying it with the given scale value (if available).
+
+### `BlockLootSubProvider`
+
+`BlockLootSubProvider` is an abstract helper class containing many helpers for creating common block loot tables, e.g. single item drops (`#createSingleItemTable`), silk touch-only drops (`#createSilkTouchOnlyTable`), drops for slab-like blocks (`#createSlabItemTable`), and many more. Unfortunately, setting up a `BlockLootSubProvider` for modded usage involves more boilerplate:
+
+```java
+public class MyBlockLootSubProvider extends BlockLootSubProvider {
+    // A list of blocks to generate tables for. Must be a mutable list because we will add values to it below.
+    private final List<Block> blocks = new ArrayList<>();
+
+    // The constructor can be private if this class is an inner class of your loot table provider.
+    public MyBlockLootSubProvider() {
+        // The first parameter is a set of blocks we are creating loot tables for. Instead of hardcoding,
+        // we use our list to dynamically add values to it and just pass an empty set here.
+        // The second parameter is the feature flag set, this will be the default flags
+        // unless you are adding custom flags (which is beyond the scope of this article).
+        super(Set.of(), FeatureFlags.DEFAULT_FLAGS);
+    }
+    
+    // We return our list here instead.
+    @Override
+    protected Iterable<Block> getKnownBlocks() {
+        return blocks;
+    }
+    
+    // We override the add() method to add the block to the list whenever we add a loot table.
+    @Override
+    protected void add(Block block, LootTable.Builder builder) {
+        super.add(block, builder);
+        blocks.add(block);
+    }
+    
+    // Actually add our loot tables.
+    @Override
+    protected void generate() {
+        add(MyBlocks.EXAMPLE_BLOCK.get(), createSingleItemTable(MyBlocks.EXAMPLE_BLOCK.get()));
+        // other loot table additions here
+    }
+}
+```
+
+We then add our sub provider to the loot table provider's constructor like any other sub provider:
+
+```java
+super(output, Set.of(), List.of(new SubProviderEntry(
+        MyBlockLootTableSubProvider::new,
+        LootContextParamSets.BLOCK // it makes sense to use BLOCK here
+)));
+```
+
+### `EntityLootSubProvider`
+
+Similar to `BlockLootSubProvider`, `EntityLootSubProvider` provides many helpers for entity loot table generation. Also similar to `BlockLootSubProvider`, we must provide an `Iterable<EntityType<?>>` of entities known to the provider. Overall, our implementation looks very similar to our `BlockLootSubProvider`, but with every mentioned of blocks swapped out for entity types:
+
+```java
+public class MyEntityLootSubProvider extends EntityLootSubProvider {
+    private final List<EntityType<?>> entities = new ArrayList<>();
```

Comment:
Again, no. Properly loop through all entities and explain the entities that are automatically excluded.